### PR TITLE
Normalize gun and ammo (parts 1)

### PR DESCRIPTION
**Pull request by mugling**
Currently the gun and ammo JSON is a mess containing many inconsistencies and errors. Property specification is obtuse and presumably confusing to new players. Examples include:
- S&W 500 magnum ammo is displayed as `damage 48`. When using the only compatible gun the actual result is `damage 56` as the gun itself is `damage 8`. This split of information makes it nearly impossible to compare ammo types adversely affecting both gameplay and balance.
- TDI vector is `reload 450` but the equivalent MP5 is `reload 100`. The majority of items are `reload 100` but the exceptions to this rule often don't make sense. Why is survivor hand cannon `reload 110`? We lack documented and enforced default values.
- `3006` ammo has `range 75` despite both player visibility being limited to 60 tiles and having no realistic chance of a hit at that distance. Many other values are also inappropriately scaled.
- `700nx` ammo stacks in 5's but the reloaded variant stacks in 30's. These definitions definitions are in separate files despite being nearly identical. Such duplication leads to errors and inconsistencies.

This PR starts an attempt to normalize the data by trying to enforce both the single responsibility principle and convention over configuration.

tldr; **specify defaults and override them only when necessary.**
#### `reload`

In 6ea2e0c we set the default of `100` and document this with example reasons for variation in d2aa127. Existing values close to `100` should be dropped in favor of this default as they have little gameplay impact and serve only to confuse. A future PR will probably rework the UI to show this and similar properties relative to the default (eg. `+300`, `-50`) along with descriptive text to help new players eg. _"much slower than usual"_
#### `aim_speed`

Aim speed is affected only by the gun (and any attached mods) and should be roughly
proportional to barrel length. 12fbfb0 provides defaults for each type of gun which are again documented in d2aa127. It is possible to override values  - for example some of the small holdout pistols now specify `aim_speed 2` to differentiate them from pistols which are by default `aim_speed 4`.
#### `range`

Range is considered to be the engagement distance as opposed to any real-world metric. It is a function of both the ammo and type of gun with SMG's and carbines improving the effective engagement distance of pistol ammo.
- Specify a default base `range` for ammo in `ammo_types.json`, eg. `9mm` is `range 7`
- If necessary provide relative adjustment to specific variants eg `9mmP` is `range +2`
- Then consider the gun type with `pistol = 0` and `smg = 4` and `rifle = 10`
- Only if unavoidable provide a relative adjustment for the gun itself

| Range with | `9mm` | `9mmP` | `9mmP2` |
| --- | --- | --- | --- |
| Glock 19 | 7 | 9 | 11 |
| H&K MP5 | 11 | 13 | 15 |
| Cx4 Storm | 17 | 19 | 21 |

As of b78d5e1 comparison between both ammo types (eg. `40` has better range than `9mm`) and  guns is now simpler and adjustments to balance become much more practical.
#### `recoil`

Is specified similar to `range`. Rifles and SMGs offset recoil by a default amount optionally with further adjustment for specific guns (b945af6). As a result firing pistol cartridges from rifles and rifle cartridges from pistols has the expected outcome without the need to specify an override:

| Recoil with | type | `223` | `556` |
| --- | --- | --- | --- |
| AR-15 | rifle | 330 | 360 |
| L2037 backup | pistol | 480 | 510 |

Again direct comparisons between different ammo is now possible. A limited number of specific override (eg. Steyr Aug, MP5) are retained where their JSON descriptions require it. Similar to `reload` if we were to express these as relative values it would help a new player pick a gun with _"less than average recoil"_
#### `MAKESHIFT`

Makeshift guns should have worse stats. Rather than specifying individual overrides the `MAKESHIFT` flag is applied which makes the required adjustments (83c01c6). A future PR will extend the UI to display the presence of such flags to the user. We can probably exploit existing flags such as `RELOAD_EJECT` and `MOUNTED_GUN` in a similar manner.
#### Future extension
- The Rivetch guns aren't normalized as I suspect @Rivet-the-Zombie will want to look into them separately. Alternatively I can undertake this. They remain entirely functional with this PR.
- Reloaded ammo should somehow become a subtype of the factory variant. Suggestions?

Please could I gently discourage anyone from threadjacking to discuss why any specific ammo should have a given property as **the goal of this PR is normalization so that all ammo has at least reasonable values**. Addressing specific balance issues is much easier after this in any case.

### CODE REVIEW

**Comment by ejseto** [comment]
>Reloaded ammo should somehow become a subtype of the factory variant. Suggestions?

Why it can't just be a flag/variable added to the item when crafted? It would be cool if crafting could apply modifiers that way, so you might have a [DELICIOUS EXTRALARGE] chocolate pancake or [RELOADED] .223 Remington or [RELOADED OVERPRESSURE] 9x19mm. There's already a [RECYCLED] flag for ammo. Rather than being an intrinsic property of an entirely different item, it should be a variable property of the same ammo, depending on how it was created. The presence of this flag could signal the game to multiply the ammo's damage and other relevant stats by, say, 95%, which is likely what happened manually during the last ammo rebalancing pass anyway.

**Comment by mugling** [comment]
@ejseto could work but fairly scary to implement.

**Comment by Coolthulhu** [comment]
I don't like the idea of hardcoding the baseline per-skill-type range. I'd rather see that as documented guideline than hardcoded bonus/penalty.

Being able to tell item's stats from the jsons alone is quite important when doing mass balancing. Hardcoded modifiers make that, and automatic listings, harder.

Also, the barrel length thing: mods should inform the player what will happen to the weapon after modding. The simplest way would probably be to create a temporary modded item and adapt `game::compare` (in inventory_ui.cpp) to show the original and the copy side by side.

**Comment by mugling** [comment]
> I don't like the idea of hardcoding the baseline per-skill-type range.

The alternative is specifying `range` manually for most items which asides from encouraging inconsistencies makes comparisons much more difficult. Presently `Cx4 Storm` and `K-SUB 2000` don't require range modifiers. If we later added `Survivor 9mm carbine` its trivial to specify `range -2` both maintaining balance and making the intention clear. In the alternative the values would be `+10, +10, +8` respectively. The problem compounds quite quickly as more things are added that influence `range`.

> Being able to tell item's stats from the jsons alone is quite important when doing mass balancing.

I'd argue it's easier. For example you scale `rifle` vs `smg` vs `pistol` range en-masse whilst maintaining the balance between weapons. More importantly you can quickly compare which weapons differ from the default values.

> Hardcoded modifiers make that, and automatic listings, harder.

The same logic counts against the `MAKESHIFT` flag and also to a per-class base `recoil`?

> Also, the barrel length thing: mods should inform the player what will happen to the weapon after modding. The simplest way would probably be to create a temporary modded item and adapt game::compare (in inventory_ui.cpp) to show the original and the copy side by side.

That's a good solution so presuming we can resolve the above I'll drop that part of this PR and go with the above implementation elsewhere

**Comment by mugling** [comment]
@Coolthulhu see d5ee25e which I'll tackle in a separate PR. Where do we go with respects to the above?

**Comment by Coolthulhu** [comment]
> Where do we go with respects to the above?

Could be what you have here, but with the ranges being explicit. Just having each rifle get ~10 range would make the weapon comparable to each other, work well with gunmods, show enough info to balance by comparison, but wouldn't require hardcoded, hidden (from the json), skill-type-based ranges.

Then add a note in balancing docs, to make sure no one decides to create range 30 rifle using a range 30 cartridge just because it is supposed to be a sniper rifle.

> The same logic counts against the MAKESHIFT flag and also to a per-class base recoil?

Yeah, I'd rather see it all in the json, unless it would mean creating a separate item type just for that one property. If MAKESHIFT allowed getting rid of all the reloaded ammo types, it could be worth it, though.

**Comment by mugling** [comment]
> If MAKESHIFT allowed getting rid of all the reloaded ammo types, it could be worth it, though

It would allow us to have one or possibly two definitions for each ammo type with modifiers `RECYCLED`, `SUBSONIC` `PIERCING` etc. This eliminates duplication which is one of our largest sources of current bugs.

> work well with gunmods

I haven't considered that. Can you expand?

> Then add a note in balancing docs, to make sure no one decides to create range 30 rifle using a range 30 cartridge just because it is supposed to be a sniper rifle

I'd actually prefer they added a `MARKSMAN` flag to the weapon which added `+5 range` and `-50 dispersion` (or whatever we decide upon). This is much less error prone and allows adjusting of all similar weapons concurrently. Similar for `MOUNTED_GUN` (which are currently unusable). It also allows the UI to extract these traits for display, eg. *This weapon is a sniper file* vs *This weapon is a machine gun*.

> show enough info to balance by comparison

Absolutely no disrespect intended to anyone but the current state of the JSON is evidence that this approach is unreliable. Different authors often disagree widely and there is a tendency for bike shedding about the semantics. In any case balancing all those items by hand is a large job and there aren't presently any suitable or willing volunteers.

I'd much prefer a traits based system as at least we can get a sane implementation in preparation for 0.D. I very much respect your opinion but I don't think we can presently achieve what your proposing. Nothing in this PR binds us in to one particular course of action - the per-gun `range` property remains in place with the same semantics so it would be trivial to drop the hardcoded value if a suitable volunteer came forward to do a manual balance.

**Comment by Coolthulhu** [comment]
> I haven't considered that. Can you expand?

For bore mods and similar mods that are more like complete conversions, the big spread of values across gun/ammo combos only causes problems. When one ammo type is balanced by normalizing values around some (non-zero) constant and another one has this value normalized to zero, the result is that one ammo type has better cartridges and other has better guns, and thus they're best used in one specific combo.

So, for example, converting a rifle to use pistol cartridges will result in tiny ranges, while converting a pistol-caliber rifle to a rifle ammo will result in huge ranges. Having all rifles have ranges normalized around non-zero value (hardcoded or otherwise) would prevent that.

> I'd actually prefer they added a MARKSMAN flag to the weapon which added +5 range and -50 dispersion (or whatever we decide upon).

That sounds like hell to manage, unless we were to get rid of vast majority of json fields here (and thus customization). Why add a flag rather than specifying the same thing explicitly?

> This is much less error prone

I disagree. Not seeing the resulting values in the json means that to view the item's stats, one has to either run the game and spawn it or calculate it from a complex formula or a sum of modifiers.

We have something like this for armors' resistances and weapons' attack speeds (though in those cases those are linear formulas and not flags). It's a pain to manage and has resulted in weird extremes in many cases.

VARSIZE flag caused a lot of number tweaking that could be all avoided by specifying the values explicitly. Though it could also be solved by inverting the order in which it is applied (making it a penalty rather than a bonus).

> Different authors often disagree widely and there is a tendency for bike shedding about the semantics.

But changing it from per-value disagreement to disagreement over entire broad generalizations sounds dangerous.

Plus it means that when we change the balance of that one flag, we either have to adjust many weapons that have it or deal with the fact that the balance is now shifted in some direction, with previously bad items getting worse or previously good items getting better.

IMO the ideal option would be to have scripts generate tables of stuff, display default values, calculate deviations from those, list outliers etc. "Softcoded formulas" for everything.

**Comment by mugling** [comment]
> Having all rifles have ranges normalized around non-zero value (hardcoded or otherwise) would prevent that

Are we in agreement a value of 10 is approximately correct?

> Why add a flag rather than specifying the same thing explicitly?

Because it's difficult to be explicit as to intention when you're specifying multiple factors. For example `MAKESHIFT` and `MARKSMAN` is explicit where as `range 6` is less so.

> `VARSIZE` flag caused a lot of number tweaking that could be all avoided by specifying the values explicitly

The main issue there was though that there are effectively two states for a `VARSIZE` item whereas this isn't the case for `MAKESHIFT` etc

> But changing it from per-value disagreement to disagreement over entire broad generalizations sounds dangerous.

Whether the algorithm is effective or not isn't affected by whether we enforce it in code or via documentation. The only difference is that the former prevents mistakes whilst still allowing manual overrides.

> one has to either run the game and spawn it or calculate it from a complex formula or a sum of modifiers.

> IMO the ideal option would be to have scripts generate tables of stuff, display default values, calculate deviations from those, list outliers etc

Such a utility would be very useful. I'd like to make some more progress with the guns so if as a compromise I write such a utility (dumping to CSV) can we call this PR done as it leaves the numbers saner than they originally were whilst retaining the option to do some manual balancing based which would be greatly eased by the usage of such a script?

**Comment by Coolthulhu** [comment]
> Are we in agreement a value of 10 is approximately correct?

Didn't look into it too much, but it sounds good.

> Because it's difficult to be explicit as to intention when you're specifying multiple factors.

Hiding them behind a tag won't make the balancing any easier.

> The main issue there was though that there are effectively two states for a VARSIZE item whereas this isn't the case for MAKESHIFT etc

The main issue was that the effect of the tag was changed during development, affecting a lot of items. The second, also significant issue, was that contributors compared the value visible in the json rather than the one in the game (which was decreased due to fitting).

> can we call this PR done

I'd much rather see that loading skill type normalization part in the jsons than the code that loads them.

And without the MAKESHIFT tag - I agree about its equivalent for ammo (where it could be used to reduce the number of similar item types), but not about the gun version, where you are unlikely to craft a "makeshift AKM".

**Comment by mugling** [comment]
> Hiding them behind a tag won't make the balancing any easier.

I don't think it makes it any harder and it's a big win from a documentation and UI perspective.

Raw values are heavily dependent upon context and it's impossible for a player to compare without implicitly knowing the rules of the algorithm. The below are poor stats on a rifle but excellent on a pistol:

`range: 12, recoil: -100 ...`
`range: 8, recoil: -150 ...`

vs

`poorer range (-2), increased recoil (+50)`
`much worse range (-2)`

(note we would display the sum value as well)

If we later decide rifles should have further increased range and adjust all the items proportionally (by an implicit constant) its impossible to assess the below stats without either the player knowing this constant or via comparison to every other gun:

`range: 12, recoil: -100 ...`

whereas the display using a coded algorithm remains the same:

`poorer range (-2), increased recoil (+50)`

If with considerable domain knowledge we struggle at times to grok the numbers then new players have effectively no chance.

> The second, also significant issue, was that contributors compared the value visible in the json rather than the one in the game

A utility to output values as CSV could also output to HTML for the wiki would adequately resolve this

**Comment by mugling** [comment]
Rebased against master as I accidentally merged in @Coolthulhu branch when I was merge testing it

**Comment by mugling** [comment]
...

> Being able to tell item's stats from the jsons alone is quite important when doing mass balancing

There are >11k lines of JSON including 557 items of type `GUN` or `AMMO`.
Paging once through these with a calculator adding values would be painful, but repeatedly?

> IMO the ideal option would be to have scripts generate tables of stuff, display default values, calculate deviations from those, list outliers etc.

The format of the JSON is immaterial with such an approach. It's objectively **much** faster to adjust `rifle +10` to `rifle +14` in code than to edit all the JSON entries by hand and also subjectively easier to specify `damage +2` for an identified outlier.

**Comment by Coolthulhu** [comment]
> It's objectively much faster to adjust rifle +10 to rifle +14 in code than to edit all the JSON entries by hand and also subjectively easier to specify damage +2 for an identified outlier.

It's a change that is unlikely to happen more than once, while small tweaks based on neighbors are going to happen. Unless this adjustment is supposed to be a temporary "I don't want to change all the jsons, let's hack this for now" thing, I'm against hardcoding the values that should be in the jsons.

It's also objectively clearer when the json displays one of the 2 factors that are added up to the final value than when it displays one of 4-5 ones, with some of them being hardcoded.

**Comment by mugling** [comment]
Please note **rebase** as I accidentally merged one of your branches during testing yesterday

8f1ae71 implements `--dump-stats` which loads the gun with the default ammo and then prints the stats for each gun in tabular (TSV) format along with any possible sawn-off variants:

`cataclysm-tiles --dump-stats GUN > output.tsv`

| Name                | ammo | vol | weight | cap | range | recoil | damage | pierce |
|---------------------|------|----|-------|----|------|------|----|----|
| AKM	           | 762 | 9	 |4690 | 30 | 19 | 370 | 36 | 8 |
| sawn-off AKM+ | 762 | 8	| 4440 | 30 | 16 | 370 | 35 | 8 |

Trivially easy to parse with a spreadsheet and then sort, filter etc to find atypical values. Conditional formatting could be very helpful here.

It displays the correct `weight` and `volume` accounting for `magazines` and `magazine_well` which are otherwise misleading in the JSON. The ammo stats are for the default ammo allowing direct comparison between items.

I'm going to extend this to dump to XML so that we can then apply an XSLT sheet to produce HTML for the wiki.

The `debugmsg` about the cerberus laser can wait for another time. Further extensions to `--dump-stats` can be a separate PR as I'd like to get this one resolved initially.

**Comment by mugling** [comment]
> Being able to tell item's stats from the jsons alone is quite important when doing mass balancing

If we need to balance frequently then adjusting algorithm constants aided by dumps is most efficient

> It's a change that is unlikely to happen more than once

If we're doing it only once then any effect on future mass rebalancing isn't relevant.

> I'm against hardcoding the values that should be in the jsons.

Definining algorithms in data is an anti-pattern. +10 for rifles followed by -1 for carbines plus any number of other (un)documented rules is an algorithm. The data is the parts that differ....

> small tweaks based on neighbors are going to happen

...such as this which does belong in JSON; eg. this rifle has atypically long range (+4). The dump tool from 8f1ae71 can be used when calculating such edits and then to verify them against the other items.

> clearer when the json displays one of the 2 factors that are added up to the final value 

`volume` appears trivial initially but is affected by `magazines`, `magazine_well` and `integral_volume`, with definitions spanning several files. It's impossible to express such semantics in data alone so you're also forced to read the source whereas the dump tool always shows the effective `volume`.

> There are >11k lines of JSON including 557 items of type GUN or AMMO.

There isn't an alternative PR using reasonable values exclusively in JSON and producing one would be a huge undertaking. In the absence of an alternative the choice is between this approach or the status quo.

> Unless this adjustment is supposed to be a temporary

Temporary in that if any future author feels strongly enough to rebalance everything en-masse then they can easily revert this PR. Permanent in the sense that's very unlikely to happen. Indeed this is the strongest argument; this PR provides both sane defaults and the tools to enable successive authors to make specific alterations as necessary. Such incremental edits much more likely to happen.

**Comment by mugling** [comment]
6bcd830 adds support for `--dump-stats AMMO`. I'm continuing work in a derived branch to implement ammo variants.

**Comment by BevapDin** [comment]
Most importantly: this PR and the proposed changes to JSON need to be reviewed by some modders (people that do JSON editing but do not write C++ code). Check whether they can (easily!) understand and use the new logic.

----

A flag from JSON that changes stats (which can also be loaded from JSON) seems bad to me:

- It's difficult to understand and to reason about (from a modders view). How does it interacts with explicitly given stats and default values. Basically what Coolthulu said. (A flag that is assigned per item instance in the game is fine.)

- Their effect is usually either to broad too narrow. A flat reduction of 5 range means long range weapon remain long range, but short range weapons may become entirely useless.

It looks like a cheap way to make things craftable: "Just copy the item definition, add the flag, and a recipe (pipe, scrap, duct tape, usual tools), no thinking needed." Requiring a proper item definition (with explicit values) encourages thinking about the data.

----

> Property specification is obtuse and presumably confusing to new players.

It seems even more obscure now. One has to look at the JSON, know what the base value is (and what it depends on) and know what values are changed by any flags (and how). Some properties have fixed default values, some have defaults based on the ammo type / the skill, some properties get modified further after loading.

Modding with explicit values seems to be easier (but I'm not a modder, therefor: see first paragraph): values from JSON  appear directly as they are in game.

----

>> This is much less error prone

> I disagree. Not seeing the resulting values in the json means that to view the item's stats, one has to either run the game and spawn it or calculate it from a complex formula or a sum of modifiers.

I agree with Coolthulhu on that.

>> I'm against hardcoding the values that should be in the jsons.

> Definining algorithms in data is an anti-pattern. +10 for rifles followed by -1 for carbines plus any number of other (un)documented rules is an algorithm. The data is the parts that differ

Do we need an algorithm with fixed values for that? AFAIK, there is no algorithm to convert a real-world value (e.g. range) to a game value, so modders would (I assume) look up what comparable items have and go from there.

Your "+10 for rifles followed by -1 for carbines plus [...]" looks like the generic gun mod.

The documentation is another *huge* factor here. It is usually outdated and incomplete, which is en even bigger problem when the values are derived by some algorithm. The current system is easy to understand: you put into JSON what should appear in the game. You want to make a new gun with a range comparable to gun X, a recoil slightly less than gun Y and a dispersion a bit more than gun Z, so you copy their respective attributes, adjust them and test. The new system makes that more difficult. (Do the other guns have explicit values for those? If not, how can I give my new gun the same effective value?)

----

In short: ask the modders what they think about this. Maybe make a thread in the modding subforum.

**Comment by kevingranade** [comment]
To be blunt, I have intentionally been keeping values in json explicit up
to this point, and I have no intention of changing that now.
I don't think convention over configuration is a reasonable principle to
use here, there is only one dda project, so there is no way for anyone to
learn the convention and then apply it later.  My focus is on keeping
barriers to entry low for new modders, and convention over configuration
does the opposite by forcing modders to learn the conventions before they
can even start contributing.

**Comment by mugling** [comment]
The consensus is against relative values so I'm going to factor out the simple JSON defaults (reload and aim speed) as separate PR's and look at ammo variants separately.
